README.md
Cheap namedtuple implementation (Python 2.7)
============================================

Namedtuples are a neat goody of Python, but they have one big caveat. In order to be
precise with the type definition, they compile with exec a string template, formated
with the variables given in the namedtuple() function factory. This compilation is very
expensive computation-wise, and in practice can be spared if needed for performance reasons.

An [issue](http://bugs.python.org/issue3974) was opened years ago in the Python bug tracker
but it got rejected, arguing that the official implementation is clearer and 
more maintainable. This is true, but there are cases where performance is key, and the current
implementation using exec is just not an option.

If you need to define new namedtuple types dynamically, and you have high performance constraints,
this is for you.

There are multiple versions out there, using [metaclasses](https://gist.github.com/aliles/1160525),
or [metaclasses](http://bugs.python.org/file11608/new_namedtuples.diff), or [ABC](http://code.activestate.com/recipes/577629-namedtupleabc-abstract-base-class-mix-in-for-named/)

This versions is simpler. Just define a new class closed-over by the factory function.

Unittests from cPython2.7 implementation are copied here to assert that the same expected behaviour is honored. 

Install
-------

```bash
pip install cheapnamedtuple
```

Usage
-----

For the purist, the `namedtuple` implementation is 3.75x faster than Python's implementation,
 while honouring 100% of the Python's implementation behaviour, including docstrings and everything else (corroborated by unit tests):

```python
>>> from cheapnamedtuple import namedtuple
>>> Point = namedtuple('Point', ['x', 'y'])
>>> Point.__doc__                   # docstring for the new class
'Point(x, y)'
>>> p = Point(11, y=22)             # instantiate with positional args or keywords
>>> p[0] + p[1]                     # indexable like a plain tuple
33
>>> x, y = p                        # unpack like a regular tuple
>>> x, y
(11, 22)
>>> p.x + p.y                       # fields also accessible by name
33
>>> d = p._asdict()                 # convert to a dictionary
>>> d['x']
11
>>> Point(**d)                      # convert from a dictionary
Point(x=11, y=22)
>>> p._replace(x=100)               # _replace() is like str.replace() but targets named fields
Point(x=100, y=22)
```

For the practical, a `cheapnamedtuple` implementation is 26.6x faster than the Python's implementation,
 while still honouring all the public behaviour (corroborated by doctests), and still supporting copy and pickle.
The only caveats identified so far are:

- The docstring of the type generated by the cheapnamedtuple factory cannot be generated ad-hoc for it
- Some typechecking has been trade in favor of performance (see test_name_fixer test case)

```python
>>> from cheapnamedtuple import cheapnamedtuple
>>> Point = cheapnamedtuple('Point', ['x', 'y'])
>>> Point.__doc__                   # docstring for the new class
'Point(x, y)'
>>> p = Point(11, y=22)             # instantiate with positional args or keywords
>>> p[0] + p[1]                     # indexable like a plain tuple
33
>>> x, y = p                        # unpack like a regular tuple
>>> x, y
(11, 22)
>>> p.x + p.y                       # fields also accessible by name
33
>>> d = p._asdict()                 # convert to a dictionary
>>> d['x']
11
>>> Point(**d)                      # convert from a dictionary
Point(x=11, y=22)
>>> p._replace(x=100)               # _replace() is like str.replace() but targets named fields
Point(x=100, y=22)
```

Compatibility
=============

Currently only tested in Python 2.7

Benchmarking
============

Python's stdlib implementation:

```bash
$python -m timeit -vvvv "from collections import namedtuple" "A = namedtuple('A', ['foo', 'bar', 'foobar'])" "a = A(1, 2, 3)" "a.bar" 
10 loops -> 0.00922394 secs
100 loops -> 0.0595999 secs
1000 loops -> 0.350676 secs
raw times: 0.328964 0.33169 0.327519
1000 loops, best of 3: 327.519 usec per loop
```

`namedtuple` implementation:

```bash
$ python -m timeit -vvvv "from cheapnamedtuple import namedtuple" "A = namedtuple('A', ['foo', 'bar', 'foobar'])" "a = A(1, 2, 3)" "a.bar"
10 loops -> 0.00332594 secs
100 loops -> 0.01106 secs
1000 loops -> 0.09164 secs
10000 loops -> 0.955008 secs
raw times: 0.929455 0.872804 0.904877
10000 loops, best of 3: 87.2804 usec per loop
```

`cheapnamedtuple` implementation:

```bash
$ python -m timeit -vvvv "from cheapnamedtuple import cheapnamedtuple" "A = cheapnamedtuple('A', ['foo', 'bar', 'foobar'])" "a = A(1, 2, 3)" "a.bar"
10 loops -> 0.00241184 secs
100 loops -> 0.00281 secs
1000 loops -> 0.0245481 secs
10000 loops -> 0.156194 secs
100000 loops -> 1.25612 secs
raw times: 1.23459 1.23159 1.23687
100000 loops, best of 3: 12.3159 usec per loop
```

Using metaclass version found [here](https://gist.github.com/aliles/1160525):

```bash
$ python -m timeit -vvvv "from metanamedtuple import namedtuple" "A = namedtuple('A', ['foo', 'bar', 'foobar'])" "a = A(1, 2, 3)" "a.bar" 
10 loops -> 0.00334907 secs
100 loops -> 0.0108609 secs
1000 loops -> 0.088969 secs
10000 loops -> 1.25756 secs
raw times: 1.2868 1.24004 1.25383
10000 loops, best of 3: 124.004 usec per loop
```


CHANGELOG.md
# CHANGELOG

## [1.1.0] 2016-07-16

### ADDED

- New implementation called 'cheapnamedtuple' for those practical of you.

## [1.0.0] 2016-07-15

- Initial release



test_cheapnamedtuple.py
"""
Verbatim copy of cpython/Lib/test/test_collections.py (2.7)
===========================================================

"""
import copy
import keyword
import pickle
import cPickle
from random import choice
import re
import string
import sys
import unittest
import doctest

import cheapnamedtuple
from cheapnamedtuple import cheapnamedtuple as namedtuple


TestNT = namedtuple('TestNT', 'x y z')    # type used for pickle tests


class TestNamedTuple(unittest.TestCase):

    def test_factory(self):
        Point = namedtuple('Point', 'x y')
        self.assertEqual(Point.__name__, 'Point')
        self.assertEqual(Point.__slots__, ())
        self.assertEqual(Point.__module__, __name__)
        self.assertEqual(Point.__getitem__, tuple.__getitem__)
        self.assertEqual(Point._fields, ('x', 'y'))

        self.assertRaises(ValueError, namedtuple, 'abc%', 'efg ghi')       # type has non-alpha char
        self.assertRaises(ValueError, namedtuple, 'class', 'efg ghi')      # type has keyword
        self.assertRaises(ValueError, namedtuple, '9abc', 'efg ghi')       # type starts with digit

        self.assertRaises(ValueError, namedtuple, 'abc', 'efg g%hi')       # field with non-alpha char
        self.assertRaises(ValueError, namedtuple, 'abc', 'abc class')      # field has keyword
        self.assertRaises(ValueError, namedtuple, 'abc', '8efg 9ghi')      # field starts with digit
        self.assertRaises(ValueError, namedtuple, 'abc', '_efg ghi')       # field with leading underscore
        self.assertRaises(ValueError, namedtuple, 'abc', 'efg efg ghi')    # duplicate field

        namedtuple('Point0', 'x1 y2')   # Verify that numbers are allowed in names
        namedtuple('_', 'a b c')        # Test leading underscores in a typename

        nt = namedtuple('nt', u'the quick brown fox')                       # check unicode input
        self.assertNotIn("u'", repr(nt._fields))
        nt = namedtuple('nt', (u'the', u'quick'))                           # check unicode input
        self.assertNotIn("u'", repr(nt._fields))

        self.assertRaises(TypeError, Point._make, [11])                     # catch too few args
        self.assertRaises(TypeError, Point._make, [11, 22, 33])             # catch too many args

    @unittest.skipIf(sys.flags.optimize >= 2,
                     "Docstrings are omitted with -O2 and above")
    def test_factory_doc_attr(self):
        Point = namedtuple('Point', 'x y')
        self.assertEqual(Point.__doc__, 'Point(x, y)')

    def test_name_fixer(self):
        for spec, renamed in [
            [('efg', 'g%hi'),  ('efg', '_1')],                              # field with non-alpha char
            [('abc', 'class'), ('abc', '_1')],                              # field has keyword
            [('8efg', '9ghi'), ('_0', '_1')],                               # field starts with digit
            [('abc', '_efg'), ('abc', '_1')],                               # field with leading underscore
            [('abc', 'efg', 'efg', 'ghi'), ('abc', 'efg', '_2', 'ghi')],    # duplicate field
            [('abc', '', 'x'), ('abc', '_1', 'x')],                         # fieldname is a space
        ]:
            self.assertEqual(namedtuple('NT', spec, rename=True)._fields, renamed)

    def test_instance(self):
        Point = namedtuple('Point', 'x y')
        p = Point(11, 22)
        self.assertEqual(p, Point(x=11, y=22))
        self.assertEqual(p, Point(11, y=22))
        self.assertEqual(p, Point(y=22, x=11))
        self.assertEqual(p, Point(*(11, 22)))
        self.assertEqual(p, Point(**dict(x=11, y=22)))
        self.assertRaises(TypeError, Point, 1)                              # too few args
        self.assertRaises(TypeError, Point, 1, 2, 3)                        # too many args
        self.assertRaises(TypeError, eval, 'Point(XXX=1, y=2)', locals())   # wrong keyword argument
        self.assertRaises(TypeError, eval, 'Point(x=1)', locals())          # missing keyword argument
        self.assertEqual(repr(p), 'Point(x=11, y=22)')
        self.assertNotIn('__weakref__', dir(p))
        self.assertEqual(p, Point._make([11, 22]))                          # test _make classmethod
        self.assertEqual(p._fields, ('x', 'y'))                             # test _fields attribute
        self.assertEqual(p._replace(x=1), (1, 22))                          # test _replace method
        self.assertEqual(p._asdict(), dict(x=11, y=22))                     # test _asdict method
        self.assertEqual(vars(p), p._asdict())                              # verify that vars() works

        try:
            p._replace(x=1, error=2)
        except ValueError:
            pass
        else:
            self._fail('Did not detect an incorrect fieldname')

        # verify that field string can have commas
        Point = namedtuple('Point', 'x, y')
        p = Point(x=11, y=22)
        self.assertEqual(repr(p), 'Point(x=11, y=22)')

        # verify that fieldspec can be a non-string sequence
        Point = namedtuple('Point', ('x', 'y'))
        p = Point(x=11, y=22)
        self.assertEqual(repr(p), 'Point(x=11, y=22)')

    def test_tupleness(self):
        Point = namedtuple('Point', 'x y')
        p = Point(11, 22)

        self.assertIsInstance(p, tuple)
        self.assertEqual(p, (11, 22))                                       # matches a real tuple
        self.assertEqual(tuple(p), (11, 22))                                # coercable to a real tuple
        self.assertEqual(list(p), [11, 22])                                 # coercable to a list
        self.assertEqual(max(p), 22)                                        # iterable
        self.assertEqual(max(*p), 22)                                       # star-able
        x, y = p
        self.assertEqual(p, (x, y))                                         # unpacks like a tuple
        self.assertEqual((p[0], p[1]), (11, 22))                            # indexable like a tuple
        self.assertRaises(IndexError, p.__getitem__, 3)

        self.assertEqual(p.x, x)
        self.assertEqual(p.y, y)
        self.assertRaises(AttributeError, eval, 'p.z', locals())

    def test_odd_sizes(self):
        Zero = namedtuple('Zero', '')
        self.assertEqual(Zero(), ())
        self.assertEqual(Zero._make([]), ())
        self.assertEqual(repr(Zero()), 'Zero()')
        self.assertEqual(Zero()._asdict(), {})
        self.assertEqual(Zero()._fields, ())

        Dot = namedtuple('Dot', 'd')
        self.assertEqual(Dot(1), (1,))
        self.assertEqual(Dot._make([1]), (1,))
        self.assertEqual(Dot(1).d, 1)
        self.assertEqual(repr(Dot(1)), 'Dot(d=1)')
        self.assertEqual(Dot(1)._asdict(), {'d':1})
        self.assertEqual(Dot(1)._replace(d=999), (999,))
        self.assertEqual(Dot(1)._fields, ('d',))

        n = 5000
        names = list(set(''.join([choice(string.ascii_letters)
                                  for j in range(10)]) for i in range(n)))
        n = len(names)
        Big = namedtuple('Big', names)
        b = Big(*range(n))
        self.assertEqual(b, tuple(range(n)))
        self.assertEqual(Big._make(range(n)), tuple(range(n)))
        for pos, name in enumerate(names):
            self.assertEqual(getattr(b, name), pos)
        repr(b)                                 # make sure repr() doesn't blow-up
        d = b._asdict()
        d_expected = dict(zip(names, range(n)))
        self.assertEqual(d, d_expected)
        b2 = b._replace(**dict([(names[1], 999),(names[-5], 42)]))
        b2_expected = range(n)
        b2_expected[1] = 999
        b2_expected[-5] = 42
        self.assertEqual(b2, tuple(b2_expected))
        self.assertEqual(b._fields, tuple(names))

    def test_pickle(self):
        p = TestNT(x=10, y=20, z=30)
        for module in pickle, cPickle:
            loads = getattr(module, 'loads')
            dumps = getattr(module, 'dumps')
            for protocol in -1, 0, 1, 2:
                q = loads(dumps(p, protocol))
                self.assertEqual(p, q)
                self.assertEqual(p._fields, q._fields)

    def test_copy(self):
        p = TestNT(x=10, y=20, z=30)
        for copier in copy.copy, copy.deepcopy:
            q = copier(p)
            self.assertEqual(p, q)
            self.assertEqual(p._fields, q._fields)


if __name__ == "__main__":
    print "Doctest:"
    print "********"
    doctest.testmod(cheapnamedtuple, verbose=True)
    print
    print "Unit test:"
    print "**********"
    unittest.main(verbosity=5)


cheapnamedtuple.py
from __future__ import print_function
from collections import Iterable, OrderedDict
from functools import partial
from keyword import iskeyword
from itertools import starmap
from operator import itemgetter
import sys

def namedtuple(typename, field_names, verbose=False, rename=False):
    """Returns a new subclass of tuple with named fields.
    >>> Point = namedtuple('Point', ['x', 'y'])
    >>> Point.__doc__                   # docstring for the new class
    'Point(x, y)'
    >>> p = Point(11, y=22)             # instantiate with positional args or keywords
    >>> p[0] + p[1]                     # indexable like a plain tuple
    33
    >>> x, y = p                        # unpack like a regular tuple
    >>> x, y
    (11, 22)
    >>> p.x + p.y                       # fields also accessible by name
    33
    >>> d = p._asdict()                 # convert to a dictionary
    >>> d['x']
    11
    >>> Point(**d)                      # convert from a dictionary
    Point(x=11, y=22)
    >>> p._replace(x=100)               # _replace() is like str.replace() but targets named fields
    Point(x=100, y=22)
    """

    # Validate the field names.  At the user's option, either generate an error
    # message or automatically replace the field name with a valid name.
    if isinstance(field_names, basestring):
        field_names = field_names.replace(',', ' ').split()
    field_names = map(str, field_names)
    typename = str(typename)
    if rename:
        seen = set()
        for index, name in enumerate(field_names):
            if (not all(c.isalnum() or c=='_' for c in name)
                or iskeyword(name)
                or not name
                or name[0].isdigit()
                or name.startswith('_')
                or name in seen):
                field_names[index] = '_%d' % index
            seen.add(name)
    for name in [typename] + field_names:
        if type(name) != str:
            raise TypeError('Type names and field names must be strings')
        if not all(c.isalnum() or c=='_' for c in name):
            raise ValueError('Type names and field names can only contain '
                             'alphanumeric characters and underscores: {}'.format(name))
        if iskeyword(name):
            raise ValueError('Type names and field names cannot be a '
                             'keyword: {}'.format(name))
        if name[0].isdigit():
            raise ValueError('Type names and field names cannot start with '
                             'a number: {}'.format(name))
    seen = set()
    for name in field_names:
        if name.startswith('_') and not rename:
            raise ValueError('Field names cannot start with an underscore: '
                             '{}'.format(name))
        if name in seen:
            raise ValueError('Encountered duplicate field name: {}'.format(name))
        seen.add(name)

    if verbose:
        print("Cheap means cheap...")

    field_names = tuple(field_names)
    arg_list = repr(tuple(field_names)).replace("'", "")[1:-1]

    class Namedtuple(tuple):
        __slots__ = ()
        _fields = tuple(field_names)

        def __new__(cls, *args, **kwargs):
            if len(args) == 1 and isinstance(args[0], Iterable):
                args = args[0]

            if len(args) > len(field_names):
                raise TypeError('Too many arguments')

            try:
                args = list(args) + [kwargs.pop(field)
                                     for field in field_names[len(args):]]
            except KeyError as e:
                raise TypeError('{} expects parameter {}'.format(typename, e))

            if kwargs:
                raise TypeError('{} got the unexpected '
                                'parameters {}'.format(typename, kwargs))

            return tuple.__new__(cls, args)

        __new__.__doc__ = \
            'Create new instance of {}({})'.format(typename, arg_list)

        def _make(cls, iterable, new=tuple.__new__, len=len):
            result = new(cls, iterable)
            if len(result) != len(field_names):
                raise TypeError(
                    'Expected {} arguments, '
                    'got {}'.format(len(field_names), len(result))
                )
            return result

        _make.__doc__ = (
            'Make a new {} object from a sequence '
            'or iterable'.format(typename)
        )

        # __doc__ of classmethods is read-only, need to wrap afterwards
        _make = classmethod(_make)

        def __repr__(self):
            'Return a nicely formatted representation string'
            return '{}({})'.format(
                typename,
                ', '.join(starmap('{}={}'.format, zip(self._fields, self)))
            )

        def _asdict(self):
            'Return a new OrderedDict which maps field names to their values'
            return OrderedDict(zip(self._fields, self))

        __dict__ = property(_asdict)

        def _replace(_self, **kwds):
            result = _self._make(map(kwds.pop, field_names, _self))
            if kwds:
                raise ValueError(
                    'Got unexpected field names: {}'.format(kwds.keys())
                )
            return result

        _replace.__doc__ = (
            'Return a new {typename} object replacing specified fields '
            'with new values'.format(typename=typename)
        )

        def __getnewargs__(self):
            'Return self as a plain tuple.  Used by copy and pickle.'
            return tuple(self)

        def __getstate__(self):
            'Exclude the OrderedDict from pickling'
            pass

        __doc__ = '{}({})'.format(typename, ', '.join(field_names))

    for index, field in enumerate(field_names):
        setattr(Namedtuple, field, property(itemgetter(index)))

    Namedtuple.__name__ = typename

    # For pickling to work, the __module__ variable needs to be set to the frame
    # where the named tuple is created.  Bypass this step in environments where
    # sys._getframe is not defined (Jython for example) or sys._getframe is not
    # defined for arguments greater than 0 (IronPython).
    try:
        Namedtuple.__module__ = sys._getframe(1).f_globals.get('__name__', '__main__')
    except (AttributeError, ValueError):
        pass

    return Namedtuple


class CheapNamedtuple(tuple):

    _fields = ()

    def __new__(cls, typename, field_names, *args, **kwargs):
        if len(args) == 1 and isinstance(args[0], Iterable):
            args = args[0]

        if len(args) > len(field_names):
            raise TypeError('Too many arguments')

        try:
            args = list(args) + [kwargs.pop(field)
                                    for field in field_names[len(args):]]
        except KeyError as e:
            raise TypeError('{} expects parameter {}'.format(typename, e))

        if kwargs:
            raise TypeError('{} got the unexpected '
                            'parameters {}'.format(typename, kwargs))

        ntuple = tuple.__new__(cls, args)
        ntuple._fields = field_names
        ntuple._typename = typename
        return ntuple

    @classmethod
    def _make(cls, typename, fields, iterable, len=len):
        return cls(typename, fields, iterable)

    def __repr__(self):
        'Return a nicely formatted representation string'
        return '{}({})'.format(
            self._typename,
            ', '.join(starmap('{}={}'.format, zip(self._fields, self)))
        )

    def _asdict(self):
        'Return a new OrderedDict which maps field names to their values'
        return OrderedDict(zip(self._fields, self))

    __dict__ = property(_asdict)

    def _replace(self, **kwds):
        result = self._make(self._typename, self._fields,
                            map(kwds.pop, self._fields, self))
        if kwds:
            raise ValueError(
                'Got unexpected field names: {}'.format(kwds.keys())
            )
        return result

    def __getnewargs__(self):
        'Return self as a plain tuple.  Used by copy and pickle.'
        return (self._typename, self._fields, tuple(self))

    def __getstate__(self):
        'Exclude the OrderedDict from pickling'
        return (self._typename, self._fields)

    def __setstate__(self, state):
        self._typename, self._fields = state


    def __getattr__(self, name):
        if name in self._fields:
            return self[self._fields.index(name)]
        raise AttributeError(name)


def cheapnamedtuple(typename, field_names, verbose=False, rename=False):
    """Returns a new subclass of tuple with named fields.
    >>> Point = cheapnamedtuple('Point', ['x', 'y'])
    >>> Point.__doc__                   # docstring for the new class
    'Point(x, y)'
    >>> p = Point(11, y=22)             # instantiate with positional args or keywords
    >>> p[0] + p[1]                     # indexable like a plain tuple
    33
    >>> x, y = p                        # unpack like a regular tuple
    >>> x, y
    (11, 22)
    >>> p.x + p.y                       # fields also accessible by name
    33
    >>> d = p._asdict()                 # convert to a dictionary
    >>> d['x']
    11
    >>> Point(**d)                      # convert from a dictionary
    Point(x=11, y=22)
    >>> p._replace(x=100)               # _replace() is like str.replace() but targets named fields
    Point(x=100, y=22)
    """

    # Validate the field names.  At the user's option, either generate an error
    # message or automatically replace the field name with a valid name.
    if isinstance(field_names, basestring):
        field_names = field_names.replace(',', ' ').split()
    field_names = tuple(map(str, field_names))
    typename = str(typename)

    adhoc_namedtuple = partial(CheapNamedtuple, typename, field_names)
    adhoc_namedtuple.__name__ = typename
    adhoc_namedtuple.__doc__ = '{}({})'.format(typename, ', '.join(field_names))
    adhoc_namedtuple._make = partial(adhoc_namedtuple.func._make,
                                     typename, field_names)
    adhoc_namedtuple._fields = field_names
    adhoc_namedtuple.__slots__ = ()
    try:
        adhoc_namedtuple.__module__ = \
            sys._getframe(1).f_globals.get('__name__', '__main__')
    except (AttributeError, ValueError):
        pass
    adhoc_namedtuple.__getitem__ = adhoc_namedtuple.func.__getitem__
    return adhoc_namedtuple
